// create-vita/dist/index.js
#!/usr/bin/env bun

// src/index.js
import { Command, Option } from "commander";
import degit from "degit";
import prompts from "prompts";

// package.json
var package_default = {
  name: "create-vita",
  description: "Scaffold a new Vite + TSX + Alpine.js project",
  type: "module",
  version: "1.1.0",
  publishConfig: {
    access: "public"
  },
  repository: {
    type: "git",
    url: "git+https://github.com/billybillydev/create-vita.git",
    directory: "packages/cli"
  },
  scripts: {
    start: "bun run dist/index.js",
    "build:watch": "tsup --watch",
    build: "tsup",
    "pub:release": "bun publish --provenance --access public"
  },
  exports: "./dist/index.js",
  bin: {
    "create-vita": "dist/index.js"
  },
  devDependencies: {
    "@types/bun": "latest",
    tsup: "^8.4.0"
  },
  peerDependencies: {
    typescript: "^5.0.0"
  },
  dependencies: {
    bun: "^1.2.7",
    commander: "^13.1.0",
    degit: "^2.8.4",
    prompts: "^2.4.2"
  }
};

// src/index.js
process.on("SIGINT", () => process.exit(0));
process.on("SIGTERM", () => process.exit(0));
async function main() {
  const program = new Command();
  program.name("create-vita").description("Scaffold a new Vite + Alpine.js project").argument("[project-name]", "name of the project folder").addOption(new Option("--with-jsxpine", "add jsxpine ui reusable components")).version(
    package_default.version || "1.0.0",
    "-v, --version",
    "display the version number"
  ).action(
    /**
     * Create a new Vite + TSX + Alpine.js project
     * @param {string} projectName name of the project folder
     * @param {object} options options
     * @param {boolean} options.withJsxpine add jsxpine ui reusable components
     */
    async (projectName, options) => {
      let withJsxpine = options?.withJsxpine ?? false;
      if (!projectName) {
        const res = await prompts({
          type: "text",
          name: "name",
          message: "Project name:",
          initial: "my-vite-alpine-app"
        });
        projectName = res.name;
      }
      if (!options?.withJsxpine) {
        const res = (
          /** @type {{ withJsxpine: boolean }} */
          await prompts({
            type: "confirm",
            name: "withJsxpine",
            message: "Add jsxpine ui reusable components?",
            initial: false
          })
        );
        withJsxpine = res.withJsxpine;
      }
      const repo = `billybillydev/create-vita/apps/${withJsxpine ? "jsxpine-" : ""}starter`;
      const emitter = degit(repo, {
        cache: false,
        force: true,
        verbose: true
      });
      await emitter.clone(projectName);
      const bunLockFile = Bun.file(
        `${process.cwd()}/${projectName}/bun.lockb`
      );
      if (await bunLockFile.exists()) {
        await bunLockFile.delete();
      }
      const pkgPath = `${process.cwd()}/${projectName}/package.json`;
      const pkgFile = Bun.file(pkgPath);
      if (await pkgFile.exists()) {
        const pkg = JSON.parse(await pkgFile.text(), "utf-8");
        pkg.name = projectName;
        await Bun.write(pkgPath, JSON.stringify(pkg, null, 2));
      }
      console.log(`
\u2705 Done!`);
      console.log(`
Next steps:
`);
      console.log(`cd ${projectName}`);
      console.log(`npm install`);
      console.log(`npm run dev
`);
    }
  );
  program.parse(process.argv);
}
main();
